sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableProvider.java

```diff
@@ -36,26 +38,47 @@
+   *
+   * @param properties The specified table properties. It's case preserving (contains exactly what
+   *                   users specified) and implementations are free to use it case sensitively or
+   *                   insensitively. It should be able to identify a table, e.g. file path, Kafka
+   *                   topic name, etc.
+   */
+  Table getTable(Map<String, String> properties);
+
+  /**
+   * Return a {@link Table} instance with specified table schema and properties to do read/write.
+   * Implementations should infer the table partitioning.
+   *
+   * @param schema The specified schema.
+   * @param properties The specified table properties. It's case preserving (contains exactly what
+   *                   users specified) and implementations are free to use it case sensitively or
+   *                   insensitively. It should be able to identify a table, e.g. file path, Kafka
+   *                   topic name, etc.
    *
-   * @param options the user-specified options that can identify a table, e.g. file path, Kafka
-   *                topic name, etc. It's an immutable case-insensitive string-to-string map.
+   * @throws IllegalArgumentException if the specified schema does not match the actual table
+   *                                  schema.
    */
-  Table getTable(CaseInsensitiveStringMap options);
+  Table getTable(StructType schema, Map<String, String> properties);
 
   /**
-   * Return a {@link Table} instance to do read/write with user-specified schema and options.
-   * <p>
-   * By default this method throws {@link UnsupportedOperationException}, implementations should
-   * override this method to handle user-specified schema.
-   * </p>
-   * @param options the user-specified options that can identify a table, e.g. file path, Kafka
-   *                topic name, etc. It's an immutable case-insensitive string-to-string map.
-   * @param schema the user-specified schema.
-   * @throws UnsupportedOperationException
+   * Return a {@link Table} instance with specified table schema, partitioning and properties to do
+   * read/write.
+   *
+   * @param schema The specified schema.
+   * @param partitioning The specified partitioning.
+   * @param properties The specified table properties. It's case preserving (contains exactly what
+   *                   users specified) and implementations are free to use it case sensitively or
+   *                   insensitively. It should be able to identify a table, e.g. file path, Kafka
+   *                   topic name, etc.
+   *
+   * @throws IllegalArgumentException if the specified schema/partitioning does not match the actual
+   *                                  table schema/partitioning.
    */
-  default Table getTable(CaseInsensitiveStringMap options, StructType schema) {
-    throw new UnsupportedOperationException(
-      this.getClass().getSimpleName() + " source does not support user-specified schema");
-  }
+  Table getTable(
```

Comment:
I am a bit curious about the parameter order in these 3 methods:
```
getTable(properties)
getTable(schema, properties)
getTable(schema, partitioning, properties)
```
Is it on purpose? Why not:
```
getTable(properties)
getTable(properties, schema)
getTable(properties, schema, partitioning)
```

Reply:
I just follow the order in `TableCatalog.createTable`

Reply:
Do you mind changing the parameter order? It is a bit wired.
Besides, previously the parameter order is like
```
getTable(options)
getTable(options, schema)
```

Reply:
I don't have a strong preference, but probably better to be consistent with `createTable`?

Reply:
Well, I think consistency in the trait TableProvider itself is more important.

Reply:
why is this not consistent?
```
getTable(properties)
getTable(schema, properties)
getTable(schema, partitioning, properties)
```

Reply:
I thought this is the comment practice. The three methods look neater when each parameter is in a fixed position
```
getTable(properties)
getTable(properties, schema)
getTable(properties, schema, partitioning)
```